gammapy/detect/cwt.py

```diff
@@ -57,200 +62,552 @@ def difference_of_gauss_kernel(radius, scale_step, n_sigmas=8):
+            log.info('Update support for scale {:.2f}'.format(self.kernels.scales[idx_scale]))
+            data.support[idx_scale] |= mask
 
-    The ``compute_support_peak`` step does the following:
+    def _inverse_transform(self, data):
+        """
+        Do the inverse transform (reconstruct the image).
+
+        TODO: describe better what this does.
+
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images for inverse transform.
+        """
+        data.mysterious_result = np.sum(data.support * data.transform, axis=0)
+        log.info('Update model')
+        data.model += data.mysterious_result * (data.mysterious_result > 0)
 
-    - computes significance ``sig = transform / error``
-    - support : 3D cube exclusion mask
-      - filled as ``sig > nsigma``
+    def _is_converged(self, data):
+        """
+        Check if the algorithm has converged on current iteration.
 
-    The ``inverse_transform`` step does the following:
+        TODO: document metric used, but not super important.
 
-    - model : 2D image, init to 0
-        - Fill ``res = np.sum(self.support * self.transform, axis=0)``
-        - Fill ``self.model += res * (res > 0)``
-        - What is this??
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Images after iteration.
 
+        Returns
+        -------
+        answer : boolean
+            Answer if CWT has converged.
+        """
+        answer = False
+        residual = data.counts - (data.model + data.approx)
+        variance = residual.var()
+        if self.previous_variance is not None:
+            variance_ratio = abs((self.previous_variance - variance) / self.previous_variance)
+            answer = variance_ratio < self.tol
+            log.info('Variance ratio: {:.7f}'.format(variance_ratio))
+        self.previous_variance = variance
+        log.info('')
+        return answer
+
+    def analyze(self, data):
+        """
+        Run iterative filter peak algorithm. The algorithm modifies the original data.
 
-    - max_scale_image : 2D image, value = index of scale where emission is most significant
+        Parameters
+        ----------
+        data : `~gammapy.detect.CWTData`
+            Input images.
+        """
+        if self.history is not None:
+            self.history = [copy.deepcopy(data)]
+
+        for n_iter in range(self.max_iter):
+            log.info('Start iteration {0}'.format(n_iter + 1))
+            self._execute_iteration(data=data)
+            if self.history is not None:
+                self.history.append(copy.deepcopy(data))
+            if self._is_converged(data=data):
+                break
+
+        if n_iter != self.max_iter:
+            log.info('Convergence reached at iteration {0}'.format(n_iter + 1))
+        else:
+            log.info('Convergence not formally reached at iteration {0}'.format(n_iter + 1))
+
+    def get_history(self):
```

Comment:
Remove `get_history`?

I don't really see how it's useful.
The user can just access the `history` attribute, and only if needed make a copy.
Usually they will not want to make a copy, but just inspect the results and avoid the duplication in required memory.